OAuth: Allow role mapping from GitHub and GitLab groups (#52407)
grafana/grafana: docs/sources/setup-grafana/configure-security/configure-authentication/github.md

allowed_organizations = github google
```

### Map roles

You can use GitHub OAuth to map roles. During mapping, Grafana checks for the presence of a role using the [JMESPath](http://jmespath.org/examples.html) specified via the `role_attribute_path` configuration option.

> For the path lookup, Grafana uses JSON obtained from querying GitHub's API [`/api/user`](https://docs.github.com/en/rest/users/users#get-the-authenticated-user=) endpoint and a `groups` key containing all of the user's teams. The result of evaluating the `role_attribute_path` JMESPath expression must be a valid Grafana role, for example, `Viewer`, `Editor` or `Admin`. For more information about roles and permissions in Grafana, refer to [Roles and permissions]({{< relref "../../../administration/roles-and-permissions/" >}}).

Reviewer: nit: I'd suggest to remove this part "and a `groups` key containing all of the user's teams", as the `group` is a bit confusing. I tried to figure out what is this groups and where it's coming from but was not able to find anything - I assume I am looking at the wrong place, however looking at the code, we just [fetch the teams](https://github.com/grafana/grafana/blob/acd85314b34bf467fbd7b4fe1d0e356ca68a1cbb/pkg/login/social/github_oauth.go#L111) and filter from there. If this is the case, I think we could just mention it.

> For the path lookup, Grafana uses JSON obtained from querying GitHub's API [`/api/user`](https://docs.github.com/en/rest/users/users#get-the-authenticated-user=) endpoint and a `groups` key containing all of the user's teams. The result of evaluating the `role_attribute_path` JMESPath expression must be a valid Grafana role, for example, `Viewer`, `Editor` or `Admin`. For more information about roles and permissions in Grafana, refer to [Roles and permissions]({{< relref "../../../administration/roles-and-permissions/" >}}).

Author: ```suggestion
For the path lookup, Grafana uses JSON obtained from querying GitHub's API [`/api/user`](https://docs.github.com/en/rest/users/users#get-the-authenticated-user=) endpoint and a `groups` key containing all of the user's teams (retrieved from `/api/user/teams`).

The result of evaluating the `role_attribute_path` JMESPath expression must be a valid Grafana role, for example, `Viewer`, `Editor` or `Admin`. For more information about roles and permissions in Grafana, refer to [Roles and permissions]({{< relref "../../../administration/roles-and-permissions/" >}}).
```

An example Query could look like the following:

```bash
role_attribute_path = [login==octocat] && 'Admin' || 'Viewer'
```

This allows the user with login "octocat" to be mapped to the `Admin` role,
but all other users to be mapped to the `Viewer` role.

#### Map roles using groups

Groups can also be used to map roles. For instance,

> if you have a group called 'example-group' you can use the following snippet to

Reviewer: nit: Could we use GitHub specific more real example, as I spent fare amount of time trying to understand why my setup is not working, until I have realized I am missing @ at the beginning. So perhaps @example-group ? Or maybe even describe with human words how are we building the shorthand version [here](https://github.com/grafana/grafana/pull/52407/files#diff-5bd5381c7a8841520acb8a969432c1454f7b5888e00d8f68ef1a62807cfde436R244).

> #### Map roles using groups
> 
> Groups can also be used to map roles. For instance,
> if you have a group called 'example-group' you can use the following snippet to
> ensure those members inherit the role 'Editor'.

Author: ```suggestion
#### Map roles using teams

Teams can also be used to map roles. For instance,
if you have a team called 'example-group' you can use the following snippet to
ensure those members inherit the role 'Editor'.
```

```bash

> role_attribute_path = contains(groups[*], 'example-group') && 'Editor' || 'Viewer'"

Reviewer: ```suggestion
role_attribute_path = contains(groups[*], 'example-group') && 'Editor' || 'Viewer'
```
